cpp/src/arrow/parquet/writer.cc

```diff
@@ -70,36 +72,124 @@ Status FileWriter::Impl::NewRowGroup(int64_t chunk_size) {
   return Status::OK();
 }
 
-template <typename ParquetType>
+#define CONVERT_PHYSICAL_TYPE_COPY_DATA(InType, OutType)                                \
+  template <>                                                                           \
+  Status FileWriter::Impl::ConvertPhysicalType<InType, OutType>(                        \
+      const InType* data_ptr, int64_t length, const OutType** out_ptr) {                \
+    RETURN_NOT_OK(data_buffer_.Resize(length * sizeof(OutType)));                       \
+    OutType* mutable_out_ptr = reinterpret_cast<OutType*>(data_buffer_.mutable_data()); \
+    std::copy(data_ptr, data_ptr + length, mutable_out_ptr);                            \
+    *out_ptr = mutable_out_ptr;                                                         \
+    return Status::OK();                                                                \
+  }
+
+#define CONVERT_PHYSICAL_TYPE_COPY_PTR(InType, OutType)           \
+  template <>                                                     \
+  Status FileWriter::Impl::ConvertPhysicalType<InType, OutType>(  \
+      const InType* data_ptr, int64_t, const OutType** out_ptr) { \
+    *out_ptr = reinterpret_cast<const OutType*>(data_ptr);        \
+    return Status::OK();                                          \
+  }
+
+CONVERT_PHYSICAL_TYPE_COPY_DATA(uint8_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int8_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(uint16_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int16_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(uint32_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(int32_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(uint64_t, int64_t)
```

Comment:
I guess I'm still thinking about Parquet 1.0. If you store the bytes in unsigned integers without a static cast, then Parquet 1.0 readers won't be able to recover the effective values:

So in your example

```
In [11]: i4_max = np.iinfo('i4').max

In [12]: u4_max = np.iinfo('u4').max

In [13]: arr = np.array([0, i4_max + 1, u4_max], dtype=np.uint32)

In [14]: arr
Out[14]: array([         0, 2147483648, 4294967295], dtype=uint32)

In [15]: arr.view(np.int32)
Out[15]: array([          0, -2147483648,          -1], dtype=int32)
```

2 out of the 3 values are lost because Parquet 1.0 readers don't know about unsigned integers.

Meanwhile, you can embed uint32 in int64 and get the correct data out:

```
In [16]: arr.astype(np.int64)
Out[16]: array([         0, 2147483648, 4294967295])
```

the compromise then is that uint32 must be stored in int64, and the other smaller unsigned integers must be stored in int32. uint64 values exceeding INT64_MAX can't be stored. In SQL databases this is the BIGINT type and so we can't handle large uint64 ints anyway.

So in summary, unlike the metadata we need to have divergent behavior around unsigned integers with Parquet 1.0 (i.e. the default right now) and 2.0